### DISCUSSION TITLE
Updating the state secret for the oauth router.

### DISCUSSION THREAD

**Discussion by xelandernt** [fastapi-users/fastapi-users#1459] [Q&A] [answered]
Hi, 
I was wondering if the `state_secret` of the `oauth_router` should be rotated for safety. Since from the current implementation it seems that it is not intended to be modified at runtime?

Is it safe to keep the secret the same for longer periods of time?
```python
app: FastAPI
fastapi_users = FastAPIUsers[User, uuid.UUID](
    get_user_manager=get_user_manager, auth_backends=[auth_backend]
)
app.include_router(
    fastapi_users.get_oauth_router(openid_auth_client, auth_backend, "STATE_SECRET"),
    prefix="/auth/openid",
    tags=["auth"],
)
```

And if it should be updated, what would be the best way to do that?

Thanks!

**Comment by frankie567** [comment]
For values like secret, the recommended way is to set them from environment variables (ref: https://12factor.net/), and not hard-code them in the code.

This way, it's safe if your source code is leaked *and* you can change it very easily by just updating the environment variable on your production server.

**Reply by xelandernt** [reply]
Thanks for the quick reply! 
Yes, however, as I mentioned, the current implementation would only read the environment variable once (at startup), so after I would update the environment variable I would need to restart the application (unless I am misunderstanding something from the FastAPI side).

**Reply by frankie567** [reply, marked as answer]
Well yes, but that's not a big deal :)

Production apps are often redeployed/restarted several times per week (or even per day). It should take only a few seconds and, if you use PaaS providers like Render or Railway, they usually handle zero-downtime deploys (i.e. the old version is shut down only when the new version is up).

---

Besides: those secrets don't have to be rotated that often. If you're really worried about it, I would say once per month is already a good cadence. But in general, if you don't suspect those secrets have leaked, you don't really need to change them.
